gateway/src/apicast/upstream.lua

```diff
@@ -0,0 +1,135 @@
+--- @classmod Upstream
+-- Abstracts how to forward traffic to upstream server.
+--- @usage
+--- local upstream = Upstream.new('http://example.com')
+--- upstream:rewrite() -- set Host header to 'example.com'
+--- -- store itself in `context` table for later use in balancer phase and call `ngx.exec`.
+--- upstream:call(context)
+
+local setmetatable = setmetatable
+local tonumber = tonumber
+local str_format = string.format
+
+local resty_resolver = require('resty.resolver')
+local resty_url = require('resty.url')
+local core_base = require('resty.core.base')
+local new_tab = core_base.new_tab
+
+local _M = {
+
+}
+
+local function proxy_pass(upstream)
+    local uri = upstream.uri
+
+    return str_format('%s://%s%s%s%s',
+            uri.scheme,
+            upstream.upstream_name,
+            uri.path or ngx.var.uri or '',
+            ngx.var.is_args or '',
+            ngx.var.query_string or '')
+end
+
+local mt = {
+    __index = _M
+}
+
+
+local function parse_url(url)
+    local parsed, err = resty_url.split(url)
+
+    if err then return nil, err end
+
+    local uri = new_tab(0, 6)
+
+    uri.scheme = parsed[1]
+    uri.user = parsed[2]
+    uri.password = parsed[3]
+    uri.host = parsed[4]
+    uri.port = tonumber(parsed[5])
+    uri.path = parsed[6]
+
+    return uri
+end
+
+
+--- Create new Upstream instance.
+--- @tparam string url
+--- @treturn Upstream|nil upstream instance
+--- @treturn nil|string error when upstream can't be initialized
+--- @static
+function _M.new(url)
+    local uri, err = parse_url(url)
+
+    if err then
+        return nil, 'invalid upstream'
+    end
+
+    return setmetatable({
+        uri = uri,
+        location_name = '@upstream',
```

Comment:
I think we should add a comment here saying that this is defined in `apicast.conf`.

Reply:
Yeo. The same for the upstream_name.